prompb/codec.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2024 Prometheus Team
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prompb
+
+import (
+	"strings"
+
+	"github.com/prometheus/common/model"
+
+	"github.com/prometheus/prometheus/model/exemplar"
+	"github.com/prometheus/prometheus/model/histogram"
+	"github.com/prometheus/prometheus/model/labels"
+)
+
+// NOTE(bwplotka): This file's code is tested in /prompb/rwcommon.
+
+// ToLabels return model labels.Labels from timeseries' remote labels.
+func (m TimeSeries) ToLabels(b *labels.ScratchBuilder, _ []string) labels.Labels {
```

Comment:
FYI @beorn7 @krajorama we cleaned up histogram conversions as it's gets tricky (duplicated) with another proto struct. Would you be happy with flow like this (see PR description).

Reply:
I guess @krajorama is way deeper into practical considerations here than I, so he should make the call.